docs/appconfiguration_generator_refact_proposal.md

```diff
@@ -0,0 +1,143 @@
+```go
+monitoringLabels := make(map[string]string)
+monitoringAnnotations := make(map[string]string)
+if g.monitoring != nil {
+    if g.project.ProjectConfiguration.Prometheus != nil && g.project.ProjectConfiguration.Prometheus.OperatorMode {
+        monitoringLabels["kusion_monitoring_appname"] = g.appName
+    } else if g.project.ProjectConfiguration.Prometheus != nil && !g.project.ProjectConfiguration.Prometheus.OperatorMode {
+        // If Prometheus doesn't run as an operator, kusion will generate the
+        // most widely-known annotation for workloads that can be consumed by
+        // the out-of-the-box community version of Prometheus server
+        // installation shown as below:
+        monitoringAnnotations["prometheus.io/scrape"] = "true"
+        monitoringAnnotations["prometheus.io/scheme"] = g.monitoring.Scheme
+        monitoringAnnotations["prometheus.io/path"] = g.monitoring.Path
+        monitoringAnnotations["prometheus.io/port"] = g.monitoring.Port
+    }
+}
+```
+**Example 2:**
+
+The `DataBaseGenerator` needs to inject env in the corresponding resources of the WorkloadGenerator. The current implementation requires the `DataBaseGenerator` to be executed before the `WorkloadGenerator`.
+```go
+func (g *databaseGenerator) injectSecret(secret *v1.Secret) error {
+    // Inject the database information into the containers of service workload.
+    if g.workload.Service != nil {
+        for _, v := range g.workload.Service.Containers {
+            v.Env[dbHostAddressEnv] = "secret://" + secret.Name + "/hostAddress"
+            v.Env[dbUsernameEnv] = "secret://" + secret.Name + "/username"
+            v.Env[dbPasswordEnv] = "secret://" + secret.Name + "/password"
+        }
+    }
+    
+    // Inject the database information into the containers of job workload.
+    if g.workload.Job != nil {
+        for _, v := range g.workload.Job.Containers {
+            v.Env[dbHostAddressEnv] = "secret://" + secret.Name + "/hostAddress"
+            v.Env[dbUsernameEnv] = "secret://" + secret.Name + "/username"
+            v.Env[dbPasswordEnv] = "secret://" + secret.Name + "/password"
+        }
+    }
+    
+    return nil
+}
+```
+## Goal
+1. Refactor the code in the Generator section to improve its extensibility, making it easier and more maintainable to add GeneratorFunc.
+
+## Non-Goal
+1.
+## Proposal
+
+We introduce a new `XXXPatcherFunc`. The current `XXXGeneratorFunc` is used only to generate new (Kubernetes) resources, while the `XXXPatcherFunc` is used only for additional patching of resources.
+```go
+func (g *appConfigurationGenerator) Generate(spec *models.Spec) error {
+	// Generator logic only generates new resources
+	gfs := []appconfiguration.NewGeneratorFunc{
+		NewNamespaceGeneratorFunc(g.project.Name),
+		accessories.NewDatabaseGeneratorFunc(g.project, g.stack, g.appName, g.app.Workload, g.app.Database),
```

Comment:
Are both workload and database needed in the new framework? I thought `DatabaseGenerator` only needed `g.app.Database` and it only generated database resources

Reply:
Yes,`g.app.Workload` will no longer needed in the new framework, I just copied current code to demonstrate the implementation order of `GeneratorFunc` and `PatcherFunc`.

I can modify it later.